drivers/flash/nrf_qspi_nor.c

```diff
@@ -519,25 +522,62 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
 
 	qspi_wait_for_completion(dev, res);
 
-	return qspi_get_zephyr_ret_code(res);
+	int rc = qspi_get_zephyr_ret_code(res);
+
+	if ((rc == 0) && (dest != dptr)) {
+		memcpy(dptr, dest, dlen);
+	}
+
+	return rc;
 }
 
 static int qspi_nor_write(struct device *dev, off_t addr, const void *src,
 			  size_t size)
 {
+	const void *sptr = src;
+	size_t dlen = size;
+	u8_t buf[4];
+
 	if (!src) {
 		return -EINVAL;
 	}
 
-	/* write size must be non-zero multiple of 4 bytes */
-	if (((size % 4U) != 0) || (size == 0)) {
-		return -EINVAL;
-	}
 	/* address must be 4-byte aligned */
 	if ((addr % 4U) != 0) {
 		return -EINVAL;
 	}
 
+	/* write size must be non-zero multiple of 4 bytes */
+	if (size < 4U) {
+		src = buf;
+		size = sizeof(buf);
+		/* read out the whole word so that unchanged data can be
+		 * written back
+		 */
+		nrfx_err_t res = nrfx_qspi_read(buf, size, addr);
+
+		if (res != 0) {
+			return qspi_get_zephyr_ret_code(res);
+		}
+		memcpy(buf, sptr, dlen);
+	} else if (((size % 4U) != 0) || (size == 0)) {
+		return -EINVAL;
+	}
+
+#if IS_ENABLED(NORDIC_QSPI_NOR_FLASH_ALLOW_STACK_USAGE_FOR_DATA_IN_FLASH)
+	if ((u32_t)src < CONFIG_SRAM_BASE_ADDRESS) {
+		/* if the value is in flash we need to copy it into RAM since
+		 * the DMA won't be able to transport data from flash to QSPI
+		 */
+		volatile u8_t rambuf[dlen];
```

Comment:
What is the change of killing stack with this? Is there a limit for `dlen`?

Reply:
No limit, I can add a Kconfig to set a limit if that sounds better?

Reply:
If you could, that would be nice.

Reply:
Since this function is blocking, could we do this in a loop with a preset buffer length? I.e. copy n bytes into buffer, then write to QSPI, repeat.

Reply:
Good point! Limits the risk that app devs accidentally let slit some big dlen value and crash the stack.

Reply:
I would assume that could work.

Reply:
@de-nordic @oyvindronningstad  could you recheck this part now.